### PR TITLE
Oauth2 built-in refresh token

### PR DESCRIPTION
**Pull request by robsontenorio**
@pi0 Another one :) 

Automatically refresh the "token" if it expires, because some Oauth providers has a short token TTL. It just intercepts axios requests, check for token expiration and try to refresh it.

Solves #148

### CODE REVIEW

**Comment by jonasgrosch** [comment]
@robsontenorio Can we get this functionality for local scheme, too?

**Comment by robsontenorio** [comment]
@jonasgrosch yes. Just waiting for approval on Oauth2 Schema.

File: lib/schemes/oauth2.js [hunk 1]
```diff
@@ -171,4 +173,71 @@ export default class Oauth2Scheme {
...
+        return config
+      }
+
+      // if already trying to refresh token, do not try again
+      if (isRefreshing) {
+        return config
+      }
+
+      // time variables
+      const tokenExpiresAt = jwtDecode(token).exp * 1000
```
Comment (ishitatsuyuki) [inline lib/schemes/oauth2.js]: Please do not rely on access token being a JWT, but instead use the `expires_in` response returned along with `access_token`.

File: lib/schemes/oauth2.js [hunk 2]
```diff
@@ -171,4 +173,71 @@ export default class Oauth2Scheme {
...
+      const refreshTokenExpiresAt = jwtDecode(refreshToken).exp * 1000
```
Comment (ishitatsuyuki) [inline lib/schemes/oauth2.js]: Please do not try to decode the refresh token. A refresh token is normally **not** a JWT, and it's the case of Auth0.

File: lib/schemes/oauth2.js [hunk 3]
```diff
@@ -171,4 +173,71 @@ export default class Oauth2Scheme {
...
+
+      // "refresh token" also has been expired. There is no way to refresh. Force logout.
+      if (now > refreshTokenExpiresAt) {
+        this.logout()
+      }
+
+      // Try to refresh token before processing current request
+      isRefreshing = true
+
+      return $axios.post(this.options.access_token_endpoint,
```
Comment (ishitatsuyuki) [inline lib/schemes/oauth2.js]: Please use async-await and when the refresh fails, please log the user out.

File: lib/schemes/oauth2.js [hunk 4]
```diff
@@ -171,4 +173,71 @@ export default class Oauth2Scheme {
...
     return true // True means a redirect happened
   }
+
+  // ---------------------------------------------------------------
+  // Watch axios requests for token expiration
+  // ---------------------------------------------------------------
+
+  watchTokenExpiration () {
+    const { $axios } = this.$auth.ctx.app
+    let isRefreshing = false
```
Comment (ishitatsuyuki) [inline lib/schemes/oauth2.js]: Please use (and wait for) a shared Promise instead of a boolean flag.

File: lib/schemes/oauth2.js (hunk 3 above)
Comment (pi0) [reply lib/schemes/oauth2.js]: I think this is safe to return a promise.

**Comment by j-evs** [comment]
Refresh token is a much needed feature. So either this or #208 would be super awesome to have!
@pi0 @ishitatsuyuki :pray::pray::pray:

**Comment by robsontenorio** [comment]
Closing in favor of #208

**Comment by HotspurHN** [comment]
any documentation of that feature? my backend returns 
```
{
    jwtToken: {
        { refreshToken: "somevalue", accessToken: "somevalue", expiresIn: "somevalue" }
    },
    response: "Success"
}
```

how to handle with that `refreshToken` here?

**Comment by robsontenorio** [comment]
It was not approved/merged, just closed. See #208

**Comment by pi0** [comment]
@robsontenorio sorry for the late review. This will be hopefully shipped with #325.

**Comment by robsontenorio** [comment]
@pi0 #325 is about LOCAL SCHEME. Will we have auto refresh token for OAUTH SCHEME?

**Comment by pi0** [comment]
@robsontenorio We may extract shared fucntionalities and share them between oauth and local.

**Comment by MathiasCiarlo** [comment]
@pi0 Could we do this and include it in 4.6.0? Or do you think it's too much work to get it done in time?

**Comment by pi0** [comment]
@MathiasCiarlo We need some refactors but don't want to block shipping other fixes for refresh support. Let's see what happens but if not included in 4.6, 4.7 or 5.0 will be worked right after that.

**Pull request closed without merging**